### PR TITLE
Make cold `bun install` use 2x less memory

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
In this benchmark: https://github.com/orogene/orogene/blob/main/BENCHMARKS.md

Under "no lockfile, no cache", this PR brings us from around 2.7 GB to 1.2 GB of memory

<details> 

<summary>package.json</summary>

```json
{
  "name": "floc",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "css": "unocss 'src/**/*.tsx'",
    "build": "npm run css && next build",
    "dev": "concurrently \"next dev\" \"npm run dev:css\"",
    "dev:css": "npm run css -- --watch",
    "postinstall": "prisma generate",
    "lint": "next lint",
    "start": "next start",
    "typecheck": "tsc -p .",
    "typecheck:watch": "npm run typecheck -- --watch"
  },
  "dependencies": {
    "@iconify-json/ph": "^1.1.4",
    "@iconify-json/uil": "^1.1.4",
    "@iconify-json/uim": "^1.1.5",
    "@iconify-json/uis": "^1.1.4",
    "@iconify-json/uit": "^1.1.4",
    "@prisma/client": "^4.11.0",
    "@react-types/shared": "^3.17.0",
    "@sindresorhus/is": "^5.3.0",
    "@tanstack/react-query": "^4.24.10",
    "@trpc/client": "^10.13.2",
    "@trpc/next": "^10.13.2",
    "@trpc/react-query": "^10.13.2",
    "@trpc/server": "^10.13.2",
    "@types/lodash-es": "^4.17.6",
    "@unocss/cli": "^0.50.3",
    "clsx": "^1.2.1",
    "concurrently": "^7.6.0",
    "eslint-plugin-simple-import-sort": "^10.0.0",
    "eslint-plugin-unused-imports": "^2.0.0",
    "husky": "^8.0.3",
    "i18next": "^22.4.10",
    "i18next-chained-backend": "^4.2.0",
    "i18next-http-backend": "^2.1.1",
    "immer": "^9.0.19",
    "iron-session": "^6.3.1",
    "lodash-es": "^4.17.21",
    "masto": "^5.10.0",
    "next": "13.2.3",
    "next-i18next": "^13.2.0",
    "react": "18.2.0",
    "react-aria": "^3.23.0",
    "react-dom": "18.2.0",
    "react-i18next": "^12.2.0",
    "react-merge-refs": "^2.0.1",
    "react-stately": "^3.21.0",
    "react-use": "^17.4.0",
    "superjson": "^1.12.2",
    "unocss": "^0.50.3",
    "zod": "^3.20.6",
    "zustand": "^4.3.5"
  },
  "devDependencies": {
    "@commitlint/cli": "^17.4.4",
    "@commitlint/config-conventional": "^17.4.4",
    "@types/node": "^18.14.2",
    "@types/prettier": "^2.7.2",
    "@types/react": "^18.0.28",
    "@types/react-dom": "^18.0.11",
    "@typescript-eslint/eslint-plugin": "^5.54.0",
    "@typescript-eslint/parser": "^5.54.0",
    "@unocss/eslint-config": "^0.50.3",
    "eslint": "^8.35.0",
    "eslint-config-next": "13.2.3",
    "lint-staged": "^13.1.2",
    "prettier": "^2.8.4",
    "prisma": "^4.11.0",
    "typescript": "^4.9.5"
  },
  "browserslist": [
    "defaults and supports es6-module"
  ],
  "ct3aMetadata": {
    "initVersion": "7.4.0"
  }
}
```

</details>

Command:
```js
rm -rf  ~/.bun/install/cache node_modules bun.lockb
gtime -v bun install
```

Will need to see what the perf impact is, as the memory allocators used here are slower at resizing large amounts of data and I haven't run this on a release build yet.

The main differences are:
- Free the package manifest API request bodies
- Free the tarballs we downloaded
- Use a smaller AST node arena (also used by the bundler) and reset it more often
- Use an arena allocator for parsing manifest JSON

This PR has no impact on the following scenario:
```bash
rm -rf  ~/.bun/install/cache node_modules
gtime -v bun install
```

Which likely means similar changes can be applied to the main thread and have a similar ~2x impact

### CODE REVIEW

File: src/install/npm.zig [hunk 1]
```diff
@@ -800,27 +807,12 @@ pub const PackageManifest = struct {
...
-            string_pool_ = String.Builder.StringPool.init(default_allocator);
-            string_pool_loaded = true;
-        }
-
-        if (!external_string_maps_loaded) {
-            external_string_maps_ = ExternalStringMapDeduper.initContext(default_allocator, .{});
-            external_string_maps_loaded = true;
-        }
-
-        if (!optional_peer_dep_names_loaded) {
```
Comment (alexlamsl) [inline src/install/npm.zig:813]: Unused `threadlocal var`:
- `string_pool_loaded`
- `external_string_maps_loaded`
- `optional_peer_dep_names_loaded`

File: src/install/npm.zig [hunk 2]
```diff
@@ -800,27 +807,12 @@ pub const PackageManifest = struct {
...
-        var optional_peer_dep_names = optional_peer_dep_names_;
-        optional_peer_dep_names.clearRetainingCapacity();
+        var string_pool = String.Builder.StringPool.init(default_allocator);
+        defer string_pool.deinit();
+        var external_string_maps = ExternalStringMapDeduper.initContext(default_allocator, .{});
+        defer external_string_maps.deinit();
+        var optional_peer_dep_names = std.ArrayList(u64).init(default_allocator);
+        defer optional_peer_dep_names.deinit();
 
         defer string_pool_ = string_pool;
```
Comment (alexlamsl) [inline src/install/npm.zig]: Assignments to unused `threadlocal var`:
- `string_pool_`
- `external_string_maps_`
- `optional_peer_dep_names_`

File: src/install/install.zig [hunk 3]
```diff
@@ -558,10 +589,18 @@ const Task = struct {
         switch (this.tag) {
             .package_manifest => {
                 var allocator = bun.default_allocator;
+                const body = this.request.package_manifest.network.response_buffer.toOwnedSliceLeaky();
+                this.request.package_manifest.network.response_buffer.list.items.len = 0;
+                this.request.package_manifest.network.response_buffer.list.capacity = 0;
```
Comment (alexlamsl) [inline src/install/install.zig]: Are these field initialisations missing before? 🤔

Comment (Jarred-Sumner) [reply src/install/install.zig]: I moved it into a method to clarify what this is doing

Basically, I want to just be really careful of any UAF caused later

**Review by alexlamsl — APPROVED** [review]
LGTM

**Pull request merged**